drivers/sensor/scd4x/scd4x.c

```diff
@@ -0,0 +1,586 @@
+/*
+ * Copyright (c) 2022 Stephen Oliver
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */   
+
+#define DT_DRV_COMPAT sensirion_scd4x
+
+#include <stdio.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/crc.h>
+
+#include <zephyr/drivers/sensor/scd4x.h>
+#include "scd4x.h"
+
+LOG_MODULE_REGISTER(SCD4X, CONFIG_SENSOR_LOG_LEVEL);
+
+
+static uint8_t scd4x_compute_crc(uint16_t value)
+{
+
+	uint8_t buf[2];
+
+	sys_put_be16(value, buf);
+
+	return crc8(buf, 2, SCD4X_CRC_POLY, SCD4X_CRC_INIT, false);
+}
+
+
+static int scd4x_write_command(const struct device *dev, uint16_t cmd)
+{
+	const struct scd4x_config *cfg = dev->config;
+	uint8_t tx_buf[2];
+
+	sys_put_be16(cmd, tx_buf);
+
+	return i2c_write_dt(&cfg->bus, tx_buf, sizeof(tx_buf));
+}
+
+
+static int scd4x_read_reg(const struct device *dev, uint16_t reg_addr, uint8_t *rx_buf, uint8_t rx_buf_size)
+{
+	const struct scd4x_config *cfg = dev->config;
+	int rc;
+
+	rc = scd4x_write_command(dev, reg_addr);
+
+	k_sleep(K_USEC(1));
+
+	rc = i2c_read_dt(&cfg->bus, rx_buf, rx_buf_size);
+
+	return rc;
+}
+
+
+static int scd4x_write_reg(const struct device *dev, uint16_t cmd, uint16_t val)
+{
+	const struct scd4x_config *cfg = dev->config;
+	uint8_t tx_buf[5];
+
+	sys_put_be16(cmd, tx_buf);
+	sys_put_be16(val, &tx_buf[2]);
+	tx_buf[4] = scd4x_compute_crc(val);
+
+	return i2c_write_dt(&cfg->bus, tx_buf, sizeof(tx_buf));
+}
+
+#if defined(SCD4X_POWER_DOWN_SINGLE_SHOT_MEASUREMENT) || defined(CONFIG_PM_DEVICE) 
+static int scd4x_power_down(const struct device *dev)
+{
+	int rc;
+
+	rc = scd4x_write_command(dev, SCD4X_CMD_POWER_DOWN);
+	k_sleep(K_MSEC(SCD4X_POWER_DOWN_WAIT_MS));
+
+	return rc;
+}
+#endif
+
+
+static void scd4x_wake_up(const struct device *dev)
+{
+	/*
+	 * The sensor does not respond to this command, regardless of whether it was successfully
+	 * received and executed or not. As a result, any error that occurs here is not detectable.
+	 */
+	scd4x_write_command(dev, SCD4X_CMD_WAKE_UP);
```

Comment:
Technically yes, datasheet version 1.2 section 3.10.4 says the sensor won't acknowledge this command at all, results is -EFAULT here on esp32. It may that the hardware can react to it and power on the rest of the sensor but is not in a state where it can ACK, or it could be that it's a no-op and that waiting the 20ms before sending a command again is what actually matters.